File: jcommander/src/main/java/xyz/cofe/jtfm/Property.java
package xyz.cofe.jtfm;

import xyz.cofe.fn.Consumer3;

/**
 * Свойство объекта
 * @param <A> Тип данных
 * @param <Self> "Собственный тип"
 */
public interface Property<A,Self extends Property<A,Self>> {
    /**
     * Получение значение свойства
     * @return значение свойства
     */
    A get();

    /**
     * Подписка на изменение свойства
     * @param listener подписчик
     * @return отписка от уведомлений
     */
    Runnable listen( Consumer3<Self,A,A> listener );
}

File: jtfm/src/main/scala/xyz/cofe/jtfm/ev/Property.scala
package xyz.cofe.jtfm.ev

/**
 * Свойство
 *
 * @tparam VALUE тип свойства
 */
trait Property[SELF <: Property[SELF, VALUE], VALUE] {
  /**
   * Чтение значения свойства
   *
   * @return значение
   */
  def value: VALUE
  
  /**
   * Обновление значение свойства
   *
   * @param v новое значение
   */
  def value(v: VALUE): Unit
  
  /**
   * Подписка на уведомления о изменении значения
   *
   * @param l подписчик
   * @return отписка от уведомлений
   */
  def listen(l: (SELF, VALUE, VALUE) => Unit): () => Unit
}
